compiler/optimizations/deadCodeElimination.cpp

```diff
@@ -437,6 +464,16 @@ static void deadBlockElimination(FnSymbol* fn)
         // then remove the entire While.
         whileStmt->remove();
 
+      else if (isSymExpr(expr))
```

Comment:
I agree that simply ignoring SymExprs is not the best solution.  It allows the PR to pass all regression tests, but it may still leave blocks or expressions in the tree that should have been removed.
The comment which follows indicates how to fix the problem: The SymExprs which have to be filtered out here should not be gathered into the list of expressions associated with that basic block (i.e. BasicBlock::exprs).  The algorithm for populating a basic block with expressions (buildBasicBlock()) was not updated correctly to deal with fields blockInfo fields that were formerly CallExprs, but are represented in the updated AST as SymExprs.  These include:
  ForLoop::mIndex
  ForLoop::mIterator
The mIndex field is converted into the condition of a for loop which means it is evaluated.  Probably, it should be converted into a primitive that is the equivalent of a statement-level symbol.  It is evaluated, but the value is not stored.  Analysis algorithms that depend on when a symbol is read should treat this as a read-access to the corresponding symbol.
I am less sure about the mIterator field.  I believe that it represents the state of the iterator.  So treating it as a statement-level symbol and evaluating it makes less sense.  It should probably be ignored when the basic block is built.
Fundamentally, if the list of expressions attached to a basic block has type std::vector<Expr*>, then code which reads this list must deal with all AST type that can be derived from Expr.  Since BlockStmt::remove() expects its operands to be CallExprs, and ForLoop inherits from LoopStmt inherits from block statement, and ForLoop does not override BlockStmt::replaceChild, something special currently be done with the SymExprs which BB analysis of ForLoop elements appends to BasicBlock::exprs.
Having done the analysis, it appears that either BasicBlock::replaceChild should be generalized to accept also a SymExpr* as its first argument, or ForLoop should override replaceChild.  In the latter approach, only ForLoop needs to know about its special SymExpr fields, so this seems like the more object-oriented approach.